### PR TITLE
Neighbour centers from `get_interp_weights` (Spin-2 preparation for furax)

### PR DESCRIPTION
**Pull request by ArtemBasyrov**
## Motivation

Interpolating Q and U as scalars is wrong. HEALPix stores them in each pixel's local meridian basis, so the four interpolation neighbours are expressed in four different bases. Combining them correctly needs a per-neighbour spin-2 rotation by the transport angle between the neighbour's center and the interpolation target.

That rotation needs each neighbour's center. `get_interp_weights` returned only `(pixels, weights)`, so a caller had to recover the centers with a second `pix2ang` pass over the four neighbours. `_get_interp_weights_ring` already computes the ring geometry those centers come from, and discarded it.

## What this adds

A `with_centers` keyword on `get_interp_weights`.

```python
pixels, weights, centers = get_interp_weights(nside, theta, phi, with_centers=True)
# centers.z    (2, *dims)  cos(theta) of the two rings
# centers.s    (2, *dims)  sin(theta) of the two rings
# centers.phi  (4, *dims)  longitude of each neighbour, aligned with `pixels`
```

`centers` is an `InterpCenters`, a `NamedTuple` and therefore a JAX pytree.

The four neighbours lie on only two rings. Rows 0 and 1 of `pixels` share the first ring and rows 2 and 3 share the second, so co-latitude takes two values per sample rather than four. Co-latitude is returned as a cosine and a sine because that is the form the ring geometry produces natively, and forming the angle would round twice.

**Centers are constants, by design.**
Pixel centers are fixed points of the HEALPix grid, so as a function of the input coordinates they are a step function. Their derivative is zero on the interior of every pixel cell. It does not exist on the cell boundaries, where the selected neighbour changes and the center jumps. Those boundaries are the ring co-latitudes in $\theta$ and the pixel edges in $\varphi$, which form a measure-zero set.

`stop_gradient` makes the zero explicit and keeps the polar `sqrt` from producing a NaN tangent at the pole. On a boundary JAX then returns zero, which is the one-sided derivative from both sides. No finite value is the derivative of a jump, so this is the same convention the existing pixel-index `stop_gradient` already uses.

Holding the centers constant is also what keeps an interpolation operator built on them linear in the map, and therefore transposable.

## Transform support

Verified: `jit`, `vmap`, `jvp`, `vjp`, `jacfwd`, `jacrev`, `grad`, `hessian`, `lax.scan`, and combinations of these.

`jax.linear_transpose` applies to an operator built from the output, not to `get_interp_weights` itself. `get_interp_weights` is not linear in `theta` and `phi`, and transposing it raises `RuntimeError`.

The operator below is linear in the map, because `pixels`, `weights` and `centers` are all constants with respect to it.

```python
pixels, weights, centers = jhp.get_interp_weights(nside, theta, phi, with_centers=True)

# spin-2 transport angle from each neighbour's centre to the target
z = centers.z[jnp.array([0, 0, 1, 1])]          # cos(theta) per neighbour
dphi = centers.phi - phi
alpha = 2.0 * jnp.arctan2(jnp.sin(dphi) * z, 1.0 + jnp.cos(dphi))
cos_a, sin_a = jnp.cos(alpha), jnp.sin(alpha)

def interpolate_qu(qu):
    """Interpolate a (2, npix) Q/U map onto the targets, rotating each neighbour."""
    q, u = qu[0][pixels], qu[1][pixels]
    return jnp.stack([
        jnp.sum(weights * (cos_a * q - sin_a * u), axis=0),
        jnp.sum(weights * (sin_a * q + cos_a * u), axis=0),
    ])

(qu_adjoint,) = jax.linear_transpose(interpolate_qu, qu)(tod)
```

Run at nside 32 in float64 with 500 targets, this satisfies $\langle A\,x,\,y\rangle = \langle x,\,A^{\mathsf{T}}y\rangle$ to a relative $3.7\times 10^{-16}$. The same identity holds to a relative $6.6\times 10^{-16}$ when the weights and centers are recomputed inside `interpolate_qu` rather than closed over. That second case is the one an iterative solver exercises.

## Implementation note

`_get_ring_costheta_sintheta` is a new helper rather than an extra return value on `_get_ring_info`. `_get_ring_info` forms cosine and sine directly only in its polar branch, and its equatorial branch goes straight to `arccos` without ever forming a sine. Both helpers run in the same jit trace on the same ring index, so common-subexpression elimination already shares the overlap. A merged prototype was tested and measured 1.756 ms against 1.740 ms for the current split, at nside 1024 over 200000 ring indices in float64, so merging would recover nothing.

## Compatibility

`with_centers` defaults to `False` and the two-value return is unchanged. Pixel selection is identical between the two settings at every nside and dtype tested. Pixels and weights are bit-identical between the two settings, at nside 16, 64, 256 and 1024 in both float32 and float64. 

That equality needs one `optimization_barrier` and does not come for free. The centers share subexpressions with the weights, and giving those nodes a second consumer changes how XLA fuses them. The co-latitude weight then divides by $\theta_2 - \theta_1 \approx 1/n_{\text{side}}$, which amplifies a one-ulp change by roughly $n_{\text{side}}$.

Without the barrier the effect is small but systematic rather than random. Over 30 seeds at each of nside 16, 64, 256 and 1024 in float64, `with_centers=True` was closer to a `healpy` reference in 0 of 120 trials, tied in 35, and further in 85, by a median factor of 1.01 to 1.15 and a worst case of 1.85. Reading the ring geometry through a barrier removes this and costs 0.25 ms of the 15.33 ms total at nside 1024 with 200000 samples in float64.

## Tests

Eleven tests in `tests/pixelfunc/test_interp.py`, 48 parametrised cases.
They cover the three branches separately, the two-rings degeneracy in the caps, shapes and ranges, float32 precision, jit, the pytree round trip, `vmap`, zero center gradients, unchanged weight gradients, and the transpose identity.

### CODE REVIEW

File: jax_healpy/pixelfunc.py [hunk 1]
```diff
@@ -1163,6 +1165,59 @@ def _get_ring_info(nside: int, ring_idx: ArrayLike) -> tuple[Array, Array, Array
...
+    elimination already shares everything that overlaps. Merging them measures no faster
+    and would give a helper that `_query_disc` also calls a variable-arity return.
+
+    Args:
+        nside (int): The healpix nside parameter.
+        ring_idx (ArrayLike): Ring index, 1 to 4*nside-1.
+
+    Returns:
+        tuple[Array, Array]: cos(theta) and sin(theta) of the ring. The sine is
+        non-negative.
```
Comment (Magwos) [inline jax_healpy/pixelfunc.py]: Cosmetic change: the style for comments in the rest of the code is more

```suggestion
    Parameters
    ----------
        nside (int): The healpix nside parameter.
        ring_idx (ArrayLike): Ring index, 1 to 4*nside-1.

    Returns
    -------
        tuple[Array, Array]: cos(theta) and sin(theta) of the ring. The sine is
        non-negative.
```

File: jax_healpy/pixelfunc.py [hunk 2]
```diff
@@ -1163,6 +1165,59 @@ def _get_ring_info(nside: int, ring_idx: ArrayLike) -> tuple[Array, Array, Array
...
+    `_get_ring_info` returns the angle, and recovering the cosine and sine from it would
+    round twice and lose accuracy near the poles, which matters in float32. It computes
+    both directly in its polar branch before collapsing them through `arctan2`, but its
+    equatorial branch goes straight to `arccos` and never forms a sine, so that half has
+    to be computed here regardless.
+
+    Kept separate rather than folded into `_get_ring_info` as an extra return value: both
+    run in the same jit trace on the same ring index, so XLA's common-subexpression
+    elimination already shares everything that overlaps. Merging them measures no faster
+    and would give a helper that `_query_disc` also calls a variable-arity return.
```
Comment (Magwos) [inline jax_healpy/pixelfunc.py]: Not necessary to keep this in the documentation I think

```suggestion
```

File: jax_healpy/pixelfunc.py [hunk 3]
```diff
@@ -1708,10 +1763,39 @@ def body(i, map_out):
...
+    Attributes:
+        z (Array): Shape (2, *dims). cos(theta) of the first and second ring.
+        s (Array): Shape (2, *dims). sin(theta) of the first and second ring,
+            non-negative.
+        phi (Array): Shape (4, *dims). Longitude of each neighbour in radians, in
+            [0, 2*pi), aligned row for row with `pixels`.
+    """
+
+    z: Array
+    s: Array
```
Comment (Magwos) [inline jax_healpy/pixelfunc.py]: Would it be possible to give them more explicit names? Maybe something like `ctheta` and `stheta`?

File: jax_healpy/pixelfunc.py [hunk 4]
```diff
@@ -1740,6 +1827,9 @@ def get_interp_weights(
     weights : Array
         Array of shape (4, N) containing the interpolation weights.
         Weights sum to 1.0 for each point to machine precision.
+    centers : InterpCenters
+        Only if with_centers is True. Centers of the pixels in `pixels`, aligned row
+        for row with it whatever order `pixels` is in.
```
Comment (Magwos) [inline jax_healpy/pixelfunc.py]: Would it be possible to specify more details on the ordering/convention of the center coordinates returned -> a bit like what you did in the PR message

> The four neighbours lie on only two rings. Rows 0 and 1 of pixels share the first ring and rows 2 and 3 share the second, so co-latitude takes two values per sample rather than four. Co-latitude is returned as a cosine and a sine because that is the form the ring geometry produces natively, and forming the angle would round twice.

File: jax_healpy/pixelfunc.py [hunk 5]
```diff
@@ -1942,7 +2036,47 @@ def _get_interp_weights_ring(nside: int, theta_coords: Array, phi_coords: Array)
     weight_sum = jnp.sum(weights, axis=0, keepdims=True)
     weights = weights / weight_sum
 
-    return pixels, weights
+    if not with_centers:
+        return pixels, weights
+
+    # Read the ring geometry through a barrier. The centers share subexpressions with the
```
Comment (Magwos) [inline jax_healpy/pixelfunc.py:2059]: Would you be able to specify in the notes of this function that a barrier is used here? (Could be useful to have it said explicitely in the documentation for the future uses)

File: jax_healpy/pixelfunc.py [hunk 6]
```diff
@@ -1942,7 +2036,47 @@ def _get_interp_weights_ring(nside: int, theta_coords: Array, phi_coords: Array)
...
+    # depend on whether centers were requested.
+    (ir1_c, ir2_c, i1_1c, i2_1c, i1_2c, i2_2c, shift1_c, shift2_c, dphi1_c, dphi2_c) = lax.optimization_barrier(
+        (ir1_safe, ir2_safe, i1_1, i2_1, i1_2, i2_2, shift1, shift2, dphi1, dphi2)
+    )
+
+    # Co-latitude of the two rings. Safe in every branch: inside a cap the clamps above
+    # collapse ir1_safe and ir2_safe onto the same ring, which is the ring the four
+    # replacement pixels lie on, so the two entries simply coincide.
+    z1, s1 = _get_ring_costheta_sintheta(nside, ir1_c)
+    z2, s2 = _get_ring_costheta_sintheta(nside, ir2_c)
```
Comment (Magwos) [inline jax_healpy/pixelfunc.py]: Same as before: if possible, would you be able to use more explicit names?

File: jax_healpy/pixelfunc.py (hunk 3 above)
Comment (ArtemBasyrov) [reply jax_healpy/pixelfunc.py]: `z` is a `healpix` convention for $\cos \theta$, but `s` is a good idea to rename to `sth` so it also follows `healpix` convention
